components/service/glean/docs/internal.md

```diff
@@ -0,0 +1,27 @@
+# Internal behavior
+
+## Clearing metrics when disabling/enabling Glean
+
+When disabling upload (`Glean.setUploadEnabled(false)`), metrics are also
+cleared to prevent their storage on the local device, and lessen the likelihood
+of accidentally sending them.  There are a couple of exceptions to this:
+
+- `first_run_date` is retained so it isn't reset if metrics are later re-enabled.
+
+- `client_id` is set to the special value
+  `"c0ffeec0-ffee-c0ff-eec0-ffeec0ffeec0"`. This should make it possible to
+  detect the error case when metrics are sent from a client that has been disabled.
+  
+When re-enabling metrics:
+
+- `first_run_date` is left as-is. (It should remain a correct date of first run
+  of the application, unaffected by disabling/enabling Glean).
+
+- `client_id` is regenerated. It has no connection to the `client_id` used prior
+  to disabling Glean.
+  
+- Application lifetime metrics owned by Glean are regenerated from scratch so
+  that they will appear in subsequent pings. This is the same process that
+  happens during every startup of the application when Glean is enabled. The
+  application is also responsible for setting any application lifetime metrics
+  that it manages at this time.
```

Comment:
Request for a follow-up bug:

If metrics are regenerated / start recording mid-application-lifecycle (after the user happened to enable), this might lead to weirdly looking data.

Could we tell when this happened vs. a new profile? Do we set any metric that allows us detecting this state?

Reply:
I don't think we do -- maybe if there's a ping with a new client id long after its first_run_date?  But that seems pretty brittle.  Anyway, I filed [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1555335)